README.md

```diff
@@ -136,31 +138,94 @@ $dbpass = Sensitive(Deferred(conjur::secret, ['production/postgres/password', {
+    ensure => file,
+    mode => '0600',
+    content => $db_password,
+  }
+
+  file { '/tmp/creds.ini':
+    ensure => file,
+    mode => '0600',
+    content => Deferred('inline_epp', [
+      'password=<%= $db_password.unwrap.unwrap %>', # **Note the use of double `unwrap`**
+      { 'db_password' => $db_password }
+    ]),
+  }
+}
+```
+
+#### `Deferred` functions
+
+This module _requires_ the use of Puppet v6+
+[`Deferred` functions](https://puppet.com/docs/puppet/6.17/deferring_functions.html)
+to ensure that the credential retrieval is fully handled on the agent. Since the
+resolution of variables is done on the agent _after_ the catalog is compiled, anything
+that requires the value of the variable during the compilation step (eg. template
+compilation)
+[must also be wrapped in a `Deferred` invocation](https://puppet.com/docs/puppet/6.17/template_with_deferred_values.html).
+
+It is also important to note that you should make sure that you invoke the
+`conjur::secret` function using the proper `Deferred` syntax:
 
-To use a secret as a string, you need to explicitly request it using the
-`unwrap` function; the result of the processing should be again wrapped in
-a `Sensitive` value.
+```puppet
+# GOOD: Passing the parameters as an array
+Sensitive(Deferred(conjur::secret, ['production/postgres/password']))
+```
+
+```puppet
+# BAD: This will not work!
+Sensitive(Deferred(conjur::secret('production/postgres/password')))
+```
 
-In particular, you should not pass unwrapped secrets as resource parameters
-if you can avoid it. Many resource types support `Sensitive` data type and
-handle it correctly. If a resource you're using does not, file a bug.
+You can read more about Puppet's `Deferred` functions
+[here](https://puppet.com/docs/puppet/6.17/deferring_functions.html).
+
+#### `Sensitive` type
+
+`conjur::secret` returns values wrapped in a `Sensitive` data type. In
+some contexts, such as string interpolation, it might cause surprising results
+(interpolating to `Sensitive [value redacted]`). This is intentional, as it
+makes it more difficult to accidentally mishandle secrets. Additionally, we also
+highly recommend that the whole `Deferred` invocation is also wrapped in
+`Sensitive` type to prevent accidental disclosure of information about the
+function parameters to the logs.
+
+To use a `Sensitive` value as a string, you need to explicitly request it using
+the `unwrap` function. In the case of the recommended usage of `conjur::secret`,
+since there are two layers of `Sensitive` wrapping, if you are using the value in
+a template, you need to use `$var_name.unwrap.unwrap` to get the raw string value.
+
+If you are setting any other variables to the value of this secret, you should
+ensure that the resulting value is also wrapped in a `Sensitive` type. In particular,
+you should not pass unwrapped secrets as resource parameters if you can avoid it. Many
+resource types support `Sensitive` data type and handle it correctly. If a
```

Comment:
I found myself needing more context when reading these 2 lines at the end of the paragraph. (Maybe the meaning would be more obvious to someone better versed in Puppet language?) More specifically, is this talking about Conjur defined resource types or more generic Puppet resource types, and if a bug is filed, what's it filed against?